### DISCUSSION TITLE
Any recommended way how to add canonical and alternate meta tags?

### DISCUSSION THREAD

**Discussion by tixyAppAdmin** [amannn/next-intl#629] [Q&A]
Seems to me that the only way to add canonical and alternate tags would be to create generateMetatags function for every page component in the app which would be a bit tedious work and create lot of boilerplate.

**Comment by amannn** [comment]
What kind of canonical and alternate meta tags are you looking for?

Note that `next-intl` already informs search engines about the different languages a given page supports:
> In addition to handling i18n routing, the middleware sets [the `link` header](https://developers.google.com/search/docs/specialty/international/localized-versions#http) to inform search engines that your content is available in different languages.

(from [the middleware docs](https://next-intl-docs.vercel.app/docs/routing/middleware))

**Reply by tixyAppAdmin** [reply]
hey, sorry for the late response. I might not be up to date with how search engines work today, but shouldn't you always have canonical and alternate tags for multi-language page? Like:

```
<link rel="canonical" href="https://abc.vercel.app/">
<link rel="alternate" hreflang="de" href="https://abc.vercel.app/de/">
<link rel="alternate" hreflang="fr" href="https://abc.vercel.app/fr/">
```


I can't see the [alternate tags](https://next-intl-docs.vercel.app/docs/routing/middleware#disable-automatic-locale-detection) being set in the [page head](https://next-intl-example-app-router.vercel.app/en)...

![image](https://github.com/amannn/next-intl/assets/135324759/324fedc8-e5ba-48b6-a4f7-130a01138246)

**Reply by amannn** [reply]
Alternate links are not set in the head, but are returned as response headers:

![Screenshot 2024-05-02 at 10 23 52](https://github.com/amannn/next-intl/assets/4038316/e4942623-bcd7-43e7-b3bb-e6af1f74f0ac)

Note that _canonical URLs should not be used_ for linking pages that are supported in different languages ([source](https://developers.google.com/search/blog/2011/12/new-markup-for-multilingual-content#annotating-pages-as-substantially-similar-content)), instead Google suggests developers to link these pages via [alternate links](https://developers.google.com/search/docs/specialty/international/localized-versions)—which is what `next-intl` does out-of-the-box.

Hope this helps!

**Reply by sovetski** [reply]
@amannn but your suggestion does not work in SSG mode right?

**Reply by amannn** [reply]
You mean `output: 'export'`? You could use a [sitemap](https://next-intl.dev/docs/environments/actions-metadata-route-handlers#sitemap) instead.

**Reply by ynamite** [reply]
> Note that _canonical URLs should not be used_ for linking pages that are supported in different languages ([source](https://developers.google.com/search/blog/2011/12/new-markup-for-multilingual-content#annotating-pages-as-substantially-similar-content)), instead Google suggests developers to link these pages via [alternate links](https://developers.google.com/search/docs/specialty/international/localized-versions)—which is what `next-intl` does out-of-the-box.
> 

@amannn while this may be true, I'm having issues getting the home page of a language indexed, because there are no canonical URLs and Google sees it as duplicate content.

I have an English version located at "/en/" and a German version located at "/de/". Google indexes the English version just fine, but won't index "/de/" and insists that it is duplicate content, when it's clearly not, and sets the canonical URL automatically to "/". While it's not a huge issue, since the default language is German, but it's still strange behavior. 

Setting language specific canonical URLs for each page in server components does not seem to be possible currently.

**Reply by amannn** [reply]
Hmm, I haven't encountered that issue so far. What's important is that Google is able to access `/de` and not being redirected away if `accept-language` doesn't match this path. If you use a default configuration for the `next-intl` middleware, that should be the case.

Is the site you're talking about this one: https://www.massif.ch? One thing I've noticed is that there are separate entries for `/en` and `/de` in the sitemap. Maybe consolidating them to one entry could help? I also found that if I search for `site:https://www.massif.ch/de/` on Google that a few pages already show up.

(Really cool website btw. 👏)

**Reply by ynamite** [reply]
That's the one :) thanks a lot, glad you like it.

Google recommends this structure for multilingual sitemaps ([source](https://developers.google.com/search/docs/specialty/international/localized-versions#sitemap)), at least as far as I understood it.

Yeah, the page is being indexed, but I just find it weird that Search Console views /de/ as duplicate content and sets the canonical URL to / 🤷 

While it probably doesn't make a big difference, it would be nice to be able to set localized canonical URL tags.

P.S. I love next-intl. Initial setup was super easy and once I got the hang of it, everything else was a breeze too! ❤️

**Reply by amannn** [reply]
A sitemap with `alternate` entries is absolutely fine, what I meant is that you have two entries that both point to your home page:

![Screenshot 2025-02-11 at 09 08 08](https://github.com/user-attachments/assets/72ba46bd-c436-47d6-b972-f81069a12f16)

**Reply by ynamite** [reply]
I know, but as far as I understood the information from Google regarding sitemaps, this is the way it should be done.
One URL points to /en/, including all alternates (even one pointing to itself).
The other points to /de/, again including all alternates (even one pointing to itself).
I'm not entirely sure if this is correct though.

**Reply by amannn** [reply]
Oh, you're totally right—I overlooked this part:

> Therefore if you have 3 versions of a page, your sitemap will have entries for the URLs of each version, and each entry will have 3 identical child entries.

Also the sample below confirms this.

Interesting, I should probably update [the sitemap docs](https://next-intl.dev/docs/environments/actions-metadata-route-handlers#sitemap) to reflect this. Many thanks for the heads-up!

**Reply by amannn** [reply]
@ynamite I just had a minute to fix the sitemap example in the docs: https://github.com/amannn/next-intl/pull/1712. I'll check again in the future if this has any effect on which pages of the App Router example are indexed on Google.

Many thanks again for bringing this up, I totally missed this when setting up the example!

**Reply by ynamite** [reply]
@amannn sure thing, glad to be able to contribute a bit!
Will be interesting to see if indexing remains unchanged or not. So long!

**Comment by ynamite** [comment]
Glad to help!
